rpcs3/util/fixed_typemap.hpp

```diff
@@ -61,138 +56,178 @@ namespace stx
-			for (auto& type : stx::typelist<typeinfo>())
+			for (const auto& type : stx::typelist<typeinfo>())
 			{
-				if (m_order[type.index()] == 0)
+				if (m_init[type.index()])
 				{
 					// Skip object if not created
-					continue;
+					_max++;
 				}
-
-				all_data[_max].object_pointer = &m_list[type.index()];
-				all_data[_max].created = m_order[type.index()];
-				all_data[_max].destroy = type.destroy;
-
-				// Clear creation order
-				m_order[type.index()] = 0;
-				_max++;
 			}
 
-			// Sort destroy list according to absolute creation order
-			destroy_info::sort_by_reverse_creation_order(all_data.get(), all_data.get() + _max);
-
-			// Destroy objects in correct order
-			for (unsigned i = 0; i < _max; i++)
+			// Destroy objects in reverse order
+			for (; _max; _max--)
 			{
-				all_data[i].destroy(*all_data[i].object_pointer);
+				(*--m_info)->destroy(*--m_order);
 			}
 
-			// Reset creation order since it now may be printed
-			m_init_count = 0;
-		}
+			// Pointers should be restored to their positions
+			delete[] m_init;
+			delete[] m_info;
+			delete[] m_order;
 
-		// Default initialize all objects if possible and not already initialized
-		void init() noexcept
-		{
-			for (auto& type : stx::typelist<typeinfo>())
+			if constexpr (Size == 0)
 			{
-				type.create(m_list[type.index()]);
-
-				// Allocate initialization order id
-				if (m_list[type.index()])
+				if (stx::typelist<typeinfo>().align() > __STDCPP_DEFAULT_NEW_ALIGNMENT__)
+				{
+					::operator delete[](m_list, std::align_val_t{stx::typelist<typeinfo>().align()});
+				}
+				else
 				{
-					m_order[type.index()] = ++m_init_count;
+					delete[] m_list;
 				}
 			}
+
+			m_init = nullptr;
+			m_info = nullptr;
+			m_order = nullptr;
+
+			if constexpr (Size == 0)
+			{
+				m_list = nullptr;
+			}
 		}
 
 		// Check if object is not initialized but shall be initialized first (to use in initializing other objects)
 		template <typename T>
 		void need() noexcept
 		{
-			if (!get<T>())
+			if (!m_init[stx::typeindex<typeinfo, std::decay_t<T>>()])
 			{
-				init<T>();
+				if constexpr (std::is_constructible_v<T, manual_typemap&>)
+				{
+					init<T>(*this);
+					return;
+				}
+
+				if constexpr (std::is_default_constructible_v<T>)
+				{
+					init<T>();
+					return;
+				}
 			}
 		}
 
-		// Explicitly (re)initialize object of type T possibly with dynamic type As and arguments
+		// Explicitly initialize object of type T possibly with dynamic type As and arguments
 		template <typename T, typename As = T, typename... Args>
 		As* init(Args&&... args) noexcept
 		{
-			auto& ptr = m_list[stx::typeindex<typeinfo, std::decay_t<T>>()];
+			if (std::exchange(m_init[stx::typeindex<typeinfo, std::decay_t<T>, std::decay_t<As>>()], true))
+			{
+				// Already exists, recreation is not supported (may be added later)
+				return nullptr;
+			}
+
+			As* obj = nullptr;
 
-			if (ptr)
+			if constexpr (Size != 0)
 			{
-				delete static_cast<T*>(ptr);
```

Comment:
Some sceNp functions use g_fxo->init<T> wrongly to spawn a thread to do work after some time, it wasnt thread-safe either so I think it's time to fix those bugs.

Reply:
cellNetCtlNetStartDialogLoadAsync

Reply:
Fixed in another PR I believe.